pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java

```diff
@@ -439,36 +372,162 @@ protected synchronized boolean trySendMessagesToConsumers(ReadType readType, Lis
             // position.
-            isDispatcherStuckOnReplays = true;
+            skipNextReplayToTriggerLookAhead = true;
             return true;
-        }  else if (currentThreadKeyNumber == 0) {
+        } else if (entriesByConsumerForDispatching.size() == 0) {
+            // There should be a backoff delay before retrying
             return true;
         }
         return false;
     }
 
-    private int getRestrictedMaxEntriesForConsumer(Consumer consumer, List<? extends Position> entries,
-           int availablePermits, ReadType readType, Set<Integer> stickyKeyHashes) {
-        int maxMessages = Math.min(entries.size(), availablePermits);
-        if (maxMessages == 0) {
-            return 0;
+    private boolean isLookAheadAllowed() {
+        if (serviceConfig.isKeySharedLookAheadEnabledWhenRecentlyJoinedConsumersExist()
+                || (recentlyJoinedConsumers == null || recentlyJoinedConsumers.isEmpty())) {
+            long keySharedNumberOfReplayMessagesThresholdForLookAhead = Math.max(
+                    serviceConfig.getKeySharedLookAheadMsgInReplayThresholdPerConsumer()
+                            * consumerList.size(),
+                    serviceConfig.getKeySharedLookAheadMsgInReplayThresholdPerSubscription());
+            if (keySharedNumberOfReplayMessagesThresholdForLookAhead == 0
+                    || redeliveryMessages.size() < keySharedNumberOfReplayMessagesThresholdForLookAhead) {
+                return true;
+            }
         }
-        if (readType == ReadType.Normal && stickyKeyHashes != null
-                && redeliveryMessages.containsStickyKeyHashes(stickyKeyHashes)) {
-            // If redeliveryMessages contains messages that correspond to the same hash as the messages
-            // that the dispatcher is trying to send, do not send those messages for order guarantee
-            return 0;
+        return false;
+    }
+
+    // groups the entries by consumer and filters out the entries that should not be dispatched
+    // the entries are handled in the order they are received instead of first grouping them by consumer and
+    // then filtering them
+    private Map<Consumer, List<Entry>> filterAndGroupEntriesForDispatching(List<Entry> entries, ReadType readType) {
+        // entries grouped by consumer
+        Map<Consumer, List<Entry>> entriesGroupedByConsumer = new HashMap<>();
+        // permits for consumer, permits are for entries/batches
+        Map<Consumer, MutableInt> permitsForConsumer = new HashMap<>();
+
+        for (Entry entry : entries) {
+            int stickyKeyHash = getStickyKeyHash(entry);
+            Consumer consumer = selector.select(stickyKeyHash);
+            boolean dispatchEntry = false;
+            // a consumer was found for the sticky key hash and the entry can be dispatched
+            if (consumer != null && canDispatchEntry(consumer, entry, readType, stickyKeyHash)) {
```

Comment:
> Addressing the consumer blocked issue: the logic has been changed to evaluate each message hash one-by-one. this is the code:

- This change has a positive effect on a normal read.
- But it has a negative effect on replay read, it also checks entries one by one.

Reply:
> But it has a negative effect on replay read, it also checks entries one by one.

@poorbarcode what is the negative effect? Checking entries one-by-one by calling a separate method isn't inefficient in this case. The JVM's JIT compiler can optimize calls and run very efficiently. It would be a different matter if IO or locks would be involved. In those cases there's a benefit of batching.

Reply:
> The JVM's JIT compiler can optimize calls and run very efficiently. It would be a different matter if IO or locks would be involved. In those cases there's a benefit of batching.

I think JIT can not optimize it very efficiently.

### 1. The method `canDispatchEntry` does a lot of things, such as follows:
- Check whether the param-position is lower than `recentlyJoinedConsumers`
- Call `removeConsumersFromRecentJoinedConsumers`

### 2. We made many improvements that save CPU resources by removing unnecessary loop
- https://github.com/apache/pulsar/pull/20607
- https://github.com/apache/pulsar/pull/19167
- https://github.com/apache/pulsar/pull/23052

Reply:
thanks for the context @poorbarcode . I see now what I need to change to optimize it.
I'll cache the result of the resolveMaxLastSentPositionForRecentlyJoinedConsumer call.

Reply:
@poorbarcode Thanks for catching the possible performance regression before it came one! Please check recent changes in 88256d6a4c8659cee759ae90851cab1ee4a1ef21, this concern should be addressed.

Reply:
I'll also address this for replays. I'm working on a larger refactoring, will push soon.

Reply:
@poorbarcode The performance regression concerns about canDispatchEntry are now addressed.